### PR TITLE
stats: Contain RawStatDataAllocator in HotRestartImpl rather than as its superclass

### PR DESCRIPTION
**Pull request by jmarantz**
*Description*: Previously, RawStatDataAllocator was not a full allocator implementation, but required alloc() and free() to be provided by HotRestartImpl, binding them together. Splitting them out makes RawStatDataAllocator easier to re-use for tests, and improves flexibility, which will help #4980 .
*Risk Level*: low -- this is a pure refactor
*Testing*: //test/...
*Docs Changes*: n/a
*Release Notes*: n/a

### CODE REVIEW

File: source/common/stats/raw_stat_data.cc [hunk 1]
```diff
@@ -35,12 +37,48 @@ uint64_t RawStatData::structSizeWithOptions(const StatsOptions& stats_options) {
 
 void RawStatData::initialize(absl::string_view key, const StatsOptions& stats_options) {
   ASSERT(!initialized());
-  ASSERT(key.size() <= stats_options.maxNameLength());
+  ASSERT(key.length() <= stats_options.maxNameLength());
```
Comment (ggreenway) [inline source/common/stats/raw_stat_data.cc]: This seems unrelated, and `size()` is used elsewhere in this PR.

File: source/common/stats/raw_stat_data.cc [hunk 2]
```diff
@@ -35,12 +37,48 @@ uint64_t RawStatData::structSizeWithOptions(const StatsOptions& stats_options) {
...
+    ++data->ref_count_;
+  }
+  return data;
+}
+
+void RawStatDataAllocator::free(Stats::RawStatData& data) {
+  // We must hold the lock since the reference decrement can race with an initialize above.
+  Thread::LockGuard lock(mutex_);
+  ASSERT(data.ref_count_ > 0);
+  if (--data.ref_count_ > 0) {
```
Comment (ggreenway) [inline source/common/stats/raw_stat_data.cc]: I know this is a copy/paste from the old version, but I intensely dislike the use of increment/decrement in the same statement as a comparison. Can you break this into two lines?

Comment (jmarantz) [reply source/common/stats/raw_stat_data.cc]: I'm kind of OK with this as it's idiomatic in C but I'm fine with establishing an envoy-wide style guide outlawing this. Out of curiosity, WDYT of the pattern:
```
  a[i++] = 5;
  a[i++] = 7;
```
should that also be banned?

In any case I changed this one.

Comment (ggreenway) [reply source/common/stats/raw_stat_data.cc]: Yes, I would be in favor of banning that pattern as well. But that's just my opinion, not the official style of the project. Feel free to leave this alone, as you just moved the lines without modification.
